src/frontends/tensorflow/src/graph_iterator_saved_model.hpp

```diff
@@ -0,0 +1,185 @@
+    /// \param path A path to file with variables data
+    /// \returns Returns true in case of everything loads successfully, false otherwise
+    bool readVariables(std::ifstream& vi_stream, const std::string& path);
+#if defined(OPENVINO_ENABLE_UNICODE_PATH_SUPPORT) && defined(_WIN32)
+    /// \brief Reads variables from opened variable index file. Can cause an asserts in case of issues.
+    /// \param vi_stream Opened stream file, file pointer doesn't matter, it will be rewind internally.
+    /// \param path A path to file with variables data
+    /// \returns Returns true in case of everything loads successfully, false otherwise
+    bool readVariables(std::ifstream& vi_stream, const std::wstring& path);
+#endif
+
+    /// \brief Returns data and size of data of stored variable
+    /// \param name Name of variable
+    /// \param data Pointer on a pointer where data pointer will be returned
+    /// \param size Pointer on a variable which will stores data size
+    /// \returns Returns true in case variable was found, false otherwise (data and size will be untouched)
+    bool getVariable(const std::string& name, char** data, size_t* size) {
+        auto varItem = varIndex.find(name);
+        if (varItem == varIndex.end()) {
+            return false;
+        }
+        *data = varItem->second.data();
+        *size = varItem->second.size();
+        return true;
+    }
+
+    /// \brief Returns data and size of data of mapped variable from trackable object graph to variables index
+    /// \param name Name of a mapping variable
+    /// \param data Pointer on a pointer where data pointer will be returned
+    /// \param size Pointer on a variable which will stores data size
+    /// \returns Returns true in case variable was found, false otherwise (data and size will be untouched)
+    bool getMappedVariable(const std::string& name, char** data, size_t* size) {
+        auto mapItem = varMap.find(name);
+        if (mapItem == varMap.end()) {
+            return false;
+        }
+        return getVariable(mapItem->second, data, size);
+    }
+
+    std::shared_ptr<std::ifstream> getDataFile(const int32_t shard_id) {
+        auto result = dataFiles.find(shard_id);
+        return result != dataFiles.end() ? result->second : nullptr;
+    }
+
+private:
+    // Internal implementation of saved model reading
+    void readVIBlock(std::ifstream& fs,
+                     const VIBlock* index,
+                     std::vector<char>& data,
+                     uint32_t* offset,
+                     uint32_t* offset_end);
+    void readVIPair(char** ptr, const char* ptr_end, std::string& key, char** value, uint32_t* val_length);
+    void readVarIndex(std::ifstream& fs, std::map<std::string, std::vector<char>>& varIndex);
+    void readBundleHeader();
+    void readCMOGraph();
+};
+
+// Loads graph from Tensorflow Saved Model file (saved_model.pb)
+class GraphIteratorSavedModel : public GraphIteratorProto {
```

Comment:
Could you explain why I shouldn’t reuse available logic from existing class? What’s your full idea of the implementation? I dont see a “final” keyword near parent class which prohibits inheritance.
Do you plan to remove support of pb models in the future? Do you plan to make a super-class which will implement everything in one place instead of separating logic?

Reply:
And later, when you will fix an issue/add feature in a one piece of code you will 99% forget to do the same in a copy.
Do you think it is wise approach?
What problem we will solve by a copying code between files?
I prefer to modify GraphIteratorProto to reuse more logic from it.

Reply:
if something is really re-used - just create helper functions for both classes, if you really like to avoid duplication

Reply:
Helper functions which I should make "friendly" for both classes? Or which will get class members as a arguments, just because of what?

Reply:
Really, I don't understand your point. At this moment GraphIteratorProto just reads pure GraphDef from a single file. GrpahIteratorSavedModel reads pure GraphDef as a member of SavedModelProto (but it still GraphDef) with some extended functionality for loading external data. I may expect to "separate" this classes only when we will find saved_model with multiple GraphDef inside, but even this might not be required too.
By using your logic: I should copy decoder_proto, make different translation list for GraphDef in SavedModel and other things. Sounds like I should copy whole project and implement "new" frontend for SavedModel support.
Why? Why I must not reuse avaiable logic? Why I cannot use OOP and inheritance? Is it common approach?

Reply:
Found in meta_graph.proto:
>     // User supplied tag(s) on the meta_graph and included graph_def.
>     //
>     // MetaGraphDefs should be tagged with their capabilities or use-cases.
>     // Examples: "train", "serve", "gpu", "tpu", etc.
>     // These tags enable loaders to access the MetaGraph(s) appropriate for a
>     // specific use-case or runtime environment.
> 

For me it means saved_model may have 1+ graphs which could be tagged to be used in different environment. Changed implementation like it described above: SavedModel reads all meta_graphs and stores all graphs. Later it could be pointed which one should be translated.